File: backend/app/core/websocket.py
from typing import Dict, List, Optional
import json
import logging
from fastapi import WebSocket, WebSocketDisconnect
from datetime import datetime

logger = logging.getLogger(__name__)

class ConnectionManager:
    """WebSocket连接管理器"""
    
    def __init__(self):
        # 存储活跃的连接：{user_id: websocket}
        self.active_connections: Dict[int, WebSocket] = {}
        # 存储用户在线状态
        self.online_users: Dict[int, Dict] = {}
    
    async def connect(self, websocket: WebSocket, user_id: int, user_info: dict):
        """建立连接"""
        await websocket.accept()
        self.active_connections[user_id] = websocket
        self.online_users[user_id] = {
            "user_id": user_id,
            "username": user_info.get("username"),
            "nickname": user_info.get("nickname"),
            "avatar": user_info.get("avatar"),
            "connected_at": datetime.now().isoformat()
        }
        
        # 通知其他用户该用户上线
        await self.broadcast_user_status(user_id, "online")
        logger.info(f"User {user_id} connected to WebSocket")
    
    async def disconnect(self, user_id: int):
        """断开连接"""
        if user_id in self.active_connections:
            del self.active_connections[user_id]
        if user_id in self.online_users:
            del self.online_users[user_id]
        
        # 通知其他用户该用户下线
        await self.broadcast_user_status(user_id, "offline")
        logger.info(f"User {user_id} disconnected from WebSocket")
    
    async def send_personal_message(self, message: dict, user_id: int):
        """发送个人消息"""
        if user_id in self.active_connections:
            try:
                await self.active_connections[user_id].send_text(json.dumps(message, ensure_ascii=False))
                return True
            except Exception as e:
                logger.error(f"Error sending message to user {user_id}: {e}")
                # 如果发送失败，移除连接
                await self.disconnect(user_id)
                return False
        return False
    
    async def send_room_message(self, message: dict, room_id: int, exclude_user: Optional[int] = None):
        """发送群聊消息到房间所有成员"""
        # 这里需要从数据库获取房间成员列表
        # 为了简化，先发送给所有在线用户（实际应用中需要查询房间成员）
        message_data = {
            "type": "room_message",
            "room_id": room_id,
            "data": message
        }
        
        failed_connections = []
        for user_id, websocket in self.active_connections.items():
            if exclude_user and user_id == exclude_user:
                continue
            try:
                await websocket.send_text(json.dumps(message_data, ensure_ascii=False))
            except Exception as e:
                logger.error(f"Error sending room message to user {user_id}: {e}")
                failed_connections.append(user_id)
        
        # 清理失败的连接
        for user_id in failed_connections:
            await self.disconnect(user_id)
    
    async def broadcast_user_status(self, user_id: int, status: str):
        """广播用户状态变化"""
        message = {
            "type": "user_status",
            "user_id": user_id,
            "status": status,
            "timestamp": datetime.now().isoformat()
        }
        
        failed_connections = []
        for uid, websocket in self.active_connections.items():
            if uid == user_id:  # 不向自己发送状态变化消息
                continue
            try:
                await websocket.send_text(json.dumps(message, ensure_ascii=False))
            except Exception as e:
                logger.error(f"Error broadcasting user status to user {uid}: {e}")
                failed_connections.append(uid)
        
        # 清理失败的连接
        for uid in failed_connections:
            await self.disconnect(uid)
    
    async def get_online_users(self) -> List[dict]:
        """获取在线用户列表"""
        return list(self.online_users.values())
    
    def is_user_online(self, user_id: int) -> bool:
        """检查用户是否在线"""
        return user_id in self.active_connections

# 全局连接管理器实例
manager = ConnectionManager() 

File: src/utils/websocket.ts
import { useUserStore } from '@/stores/user'

export interface WebSocketMessage {
  type: string
  data: any
}

interface ChatMessage {
  id: number
  sender_id: number
  sender_username: string
  sender_nickname: string
  sender_avatar?: string
  receiver_id?: number
  room_id?: number
  content: string
  message_type: string
  created_at: string
  is_read: boolean
}

export class WebSocketManager {
  private ws: WebSocket | null = null
  private reconnectAttempts = 0
  private maxReconnectAttempts = 5
  private reconnectInterval = 3000
  private messageHandlers: Map<string, (data: any) => void> = new Map()
  private onlineUsers: any[] = []
  
  constructor() {
    // 绑定方法
    this.connect = this.connect.bind(this)
    this.disconnect = this.disconnect.bind(this)
    this.sendMessage = this.sendMessage.bind(this)
    this.addMessageHandler = this.addMessageHandler.bind(this)
    this.removeMessageHandler = this.removeMessageHandler.bind(this)
  }

  async connect(): Promise<boolean> {
    try {
      const userStore = useUserStore()
      if (!userStore.token) {
        console.error('No token available for WebSocket connection')
        return false
      }

      // 构造WebSocket URL
      const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
      const host = import.meta.env.VITE_API_BASE_URL?.replace(/^https?:\/\//, '') || 'localhost:8000'
      const wsUrl = `${protocol}//${host}/api/v1/ws/${userStore.token}`

      this.ws = new WebSocket(wsUrl)

      this.ws.onopen = () => {
        console.log('WebSocket connected')
        this.reconnectAttempts = 0
      }

      this.ws.onmessage = (event) => {
        try {
          const message: WebSocketMessage = JSON.parse(event.data)
          this.handleMessage(message)
        } catch (error) {
          console.error('Error parsing WebSocket message:', error)
        }
      }

      this.ws.onclose = (event) => {
        console.log('WebSocket disconnected:', event.code, event.reason)
        this.ws = null
        
        // 自动重连
        if (this.reconnectAttempts < this.maxReconnectAttempts) {
          setTimeout(() => {
            this.reconnectAttempts++
            console.log(`Attempting to reconnect... (${this.reconnectAttempts}/${this.maxReconnectAttempts})`)
            this.connect()
          }, this.reconnectInterval)
        }
      }

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error)
      }

      return true
    } catch (error) {
      console.error('Failed to connect to WebSocket:', error)
      return false
    }
  }

  disconnect() {
    if (this.ws) {
      this.ws.close()
      this.ws = null
    }
  }

  sendMessage(message: WebSocketMessage): boolean {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      try {
        this.ws.send(JSON.stringify(message))
        return true
      } catch (error) {
        console.error('Error sending WebSocket message:', error)
        return false
      }
    }
    console.warn('WebSocket is not connected')
    return false
  }

  private handleMessage(message: WebSocketMessage) {
    const handler = this.messageHandlers.get(message.type)
    if (handler) {
      handler(message.data)
    } else {
      console.log('Unhandled message type:', message.type, message.data)
    }

    // 处理内置消息类型
    switch (message.type) {
      case 'online_users':
        this.onlineUsers = message.data
        break
      case 'user_status':
        this.handleUserStatus(message.data)
        break
    }
  }

  private handleUserStatus(data: any) {
    const { user_id, status } = data
    if (status === 'online') {
      // 添加到在线用户列表
      if (!this.onlineUsers.find(user => user.user_id === user_id)) {
        this.onlineUsers.push({
          user_id,
          username: data.username,
          nickname: data.nickname,
          avatar: data.avatar,
          connected_at: data.timestamp
        })
      }
    } else if (status === 'offline') {
      // 从在线用户列表移除
      this.onlineUsers = this.onlineUsers.filter(user => user.user_id !== user_id)
    }
  }

  addMessageHandler(type: string, handler: (data: any) => void) {
    this.messageHandlers.set(type, handler)
  }

  removeMessageHandler(type: string) {
    this.messageHandlers.delete(type)
  }

  // 发送私聊消息
  sendPrivateMessage(receiverId: number, content: string, messageType: string = 'text'): boolean {
    return this.sendMessage({
      type: 'private_message',
      data: {
        receiver_id: receiverId,
        content,
        message_type: messageType
      }
    })
  }

  // 发送群聊消息
  sendRoomMessage(roomId: number, content: string, messageType: string = 'text'): boolean {
    return this.sendMessage({
      type: 'room_message',
      data: {
        room_id: roomId,
        content,
        message_type: messageType
      }
    })
  }

  // 发送正在输入状态
  sendTypingStatus(targetId: number, targetType: 'user' | 'room', isTyping: boolean): boolean {
    return this.sendMessage({
      type: 'typing',
      data: {
        target_id: targetId,
        target_type: targetType,
        is_typing: isTyping
      }
    })
  }

  // 标记消息已读
  markMessageAsRead(messageId: number): boolean {
    return this.sendMessage({
      type: 'read_message',
      data: {
        message_id: messageId
      }
    })
  }

  // 获取在线用户列表
  getOnlineUsers(): any[] {
    return this.onlineUsers
  }

  // 检查用户是否在线
  isUserOnline(userId: number): boolean {
    return this.onlineUsers.some(user => user.user_id === userId)
  }

  // 获取连接状态
  isConnected(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN
  }
}

// 全局WebSocket管理器实例
export const wsManager = new WebSocketManager() 